client/src/main/java/io/cloudsoft/winrm4j/client/WinRmClient.java

```
@@ -244,41 +252,62 @@ private int receiveCommand(String commandId, Writer out, Writer err) {
-                if (out != null && "stdout".equals(s.getName())) {
-                    try {
-                        //TODO use passed locale?
-                        if (value.length > 0) {
-                            out.write(new String(value));
-                        }
-                        if (Boolean.TRUE.equals(s.isEnd())) {
-                            out.close();
-                        }
-                    } catch (IOException e) {
-                        throw new IllegalStateException(e);
-                    }
+
+            try {
+                numberOfReceiveCalls++;
+                ReceiveResponse receiveResponse = winrm.receive(receive, RESOURCE_URI, MAX_ENVELOPER_SIZE, operationTimeout, locale, shellSelector);
+                getStreams(receiveResponse, out, err);
+
+                System.out.println(String.format("receiveResponse: %s", receiveResponse.toString()));
+                CommandStateType state = receiveResponse.getCommandState();
+                if (COMMAND_STATE_DONE.equals(state.getState())) {
+                    return state.getExitCode().intValue();
                 }
-                if (err != null && "stderr".equals(s.getName())) {
-                    try {
-                        //TODO use passed locale?
-                        if (value.length > 0) {
-                            err.write(new String(value));
-                        }
-                        if (Boolean.TRUE.equals(s.isEnd())) {
-                            err.close();
-                        }
-                    } catch (IOException e) {
-                        throw new IllegalStateException(e);
+            } catch (SOAPFaultException e) {
+                // e.getMessage().contains("The WS-Management service cannot complete the operation within the time specified in OperationTimeout")
+                if(e.getFault().getDetail().getFirstChild().getAttributes().getNamedItem("Code").getNodeValue().equals("2150858793")) {
```

I completely agree it should throw the previous exception, hence the need for handling the nulls.

I'd prefer an explicit check rather than wrapping in a try-catch, to handle nulls. It's a little verbose but generally preferable to handle nulls with a check rather than a try-catch.